### PR TITLE
[dtensor] fix new_empty_strided op

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107835

This PR fixes the new_empty_strided op to become replicate from sharding
when necessary, this is a quick fix to resolve https://github.com/pytorch/pytorch/issues/107661

We'll need to think more about the behavior of this op when it comes to
sharding, one possibility is to follow the input sharding, but given the
output shape of this op might not be the same as the input, it's hard to
say we should follow the input sharding, further improvement needed once
we figure out the op syntax

### CODE REVIEW

File: test/distributed/_tensor/test_dtensor.py [hunk 1]
```diff
@@ -245,6 +245,22 @@ def test_to_local(self):
         except RuntimeError:
             self.assertEqual(sharded_tensor.grad.stride(), [1, 3 * self.world_size])
 
+    @with_comms
+    def test_dtensor_new_empty_strided(self):
+        device_mesh = DeviceMesh(self.device_type, torch.arange(self.world_size))
+        local_tensor = torch.randn(8, 8, requires_grad=True, device=self.device_type)
+        my_dtensor = distribute_tensor(local_tensor, device_mesh, [Shard(0)])
+        new_strided_dtensor = my_dtensor.new_empty_strided((8, 8), (8, 1))
+        self.assertEqual(new_strided_dtensor.shape, my_dtensor.shape)
```
Comment (fduwjj) [inline test/distributed/_tensor/test_dtensor.py:254]: Any reason why we don't test BWD?

Comment (wanchaol) [reply test/distributed/_tensor/test_dtensor.py:254]: you mean after new_empty_strided then sum().backward? I can write that test too

File: test/distributed/_tensor/test_dtensor.py [hunk 2]
```diff
@@ -245,6 +245,28 @@ def test_to_local(self):
...
+        self.assertIsNotNone(new_strided_dtensor.grad)
+        self.assertIsInstance(new_strided_dtensor.grad, DTensor)
+
+        # test backward new_empty_strided with sharding works correctly
+        my_dtensor.to_local().sum().backward()
+        local_tensor.sum().backward()
+        self.assertEqual(
+            my_dtensor.grad.redistribute(placements=[Replicate()]).to_local(),
+            local_tensor.grad,
+        )
```
Comment (fduwjj) [inline test/distributed/_tensor/test_dtensor.py:268]: Do we also want to compare `new_strided_dtensor.grad` and `my_dtensor.grad` here?

**Review by fduwjj — APPROVED** [review]
LGTM and one nit in the test.

File: test/distributed/_tensor/test_dtensor.py (hunk 2 above)
Comment (wanchaol) [reply test/distributed/_tensor/test_dtensor.py:269]: done :)

**Pull request closed without merging**